hal: AArch64/ZynqMP port

jmaksymowicz opened the pull request:
Port plo to AArch64 architecture and Zynq Ultrascale+ platform.

## Description
<!--- Describe your changes shortly -->

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Chore (refactoring, style fixes, git/CI config, submodule management, no code logic changes)

<!--- In case of breaking change - please advice here what needs to be done in dependent projects. -->

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
- [ ] Already covered by automatic testing.
- [ ] New test added: (add PR link here).
- [x] Tested by hand on: aarch64a53-zynqmp-qemu, aarch64a53-zynqmp-zcu104

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [x] All new and existing linter checks and tests passed.
- [x] My changes generate no new compilation warnings for any of the targets.

## Special treatment

- [ ] This PR needs additional PRs to work (list the PRs, preferably in merge-order).
- [ ] I will merge this PR by myself when appropriate.

jmaksymowicz commented on hal/aarch64/exceptions.c:
```diff
@@ -0,0 +1,216 @@
+		case 0b000110:
+			return "Trapped LDC/STC access";
+		case 0b000111:
+			return "Trapped SME, SVE, Advanced SIMD or floating-point functionality due to CPACR_ELx.FPEN";
```
Not a typo: SME is Scalable Matrix Extensions

jmaksymowicz commented on hal/aarch64/exceptions.c:
```diff
@@ -0,0 +1,216 @@
+		case 0b011100:
+			return "(FEAT_FPAC) Exception from a PAC Fail";
+		case 0b011101:
+			return "(FEAT_SME) Access to SME functionality trapped";
```
Not a typo: SME is Scalable Matrix Extensions

agkaminski replied:
Add to .codespell_ignore?

jmaksymowicz commented on devices/gpio-zynq7000/gpio.c:
```diff
@@ -18,28 +18,48 @@
+	data_lsw = 0, data_msw, /* up to MAX_BANK_NB pairs of 2 registers*/
+	data = 0x10, /* up to MAX_BANK_NB registers */
+	data_ro = 0x18, /* up to MAX_BANK_NB registers */
+	dirm = 0x81, oen, int_mask, int_en, int_dis, int_stat, int_type, int_pol, int_any, /* MAX_BANK_NB banks, BANK_REGS_OFFS stride */
```
Not a typo: stands for "**o**utput **en**able"

jmaksymowicz commented on hal/aarch64/zynqmp/_init.S:
```diff
@@ -0,0 +1,308 @@
+	ldr x0, =0x30c50838
+	msr sctlr_el3, x0
+
+	/* Note: setting to a value of 0 will trap execution of SME instructions */
```
Not a typo: SME is Scalable Matrix Extensions

agkaminski commented on hal/aarch64/Makefile:
```diff
@@ -0,0 +1,14 @@
+#
+# Makefile for Phoenix-RTOS loader (AArch64 HAL)
+#
+# Copyright 2021 Phoenix Systems
```
```suggestion
# Copyright 2025 Phoenix Systems
```

agkaminski commented on hal/aarch64/_exceptions.S:
```diff
@@ -0,0 +1,74 @@
+ *
+ * Exceptions handlers for Cortex-A v. 7
+ *
+ * Copyright 2018, 2020, 2021 Phoenix Systems
+ * Author: Pawel Pisarczyk, Aleksander Kaminski, Maciej Purski, Hubert Buczynski
```
Update pls

agkaminski commented on hal/aarch64/_interrupts.S:
```diff
@@ -0,0 +1,66 @@
+ *
+ * Operating system loader
+ *
+ * Interrupts handlers for Cortex-A v. 7
```
Doesn't seem right

agkaminski commented on hal/aarch64/_exceptions.S:
```diff
@@ -0,0 +1,74 @@
+ *
+ * Operating system loader
+ *
+ * Exceptions handlers for Cortex-A v. 7
```
Doesn't seem right

agkaminski commented on hal/aarch64/mmu.c:
```diff
@@ -0,0 +1,169 @@
+#define DESCR_ATTR(x) (((x) & 0x7) << 2) /* Memory attribute from MAIR_EL1 */
+#define DESCR_AP1     (1uL << 6)         /* Unprivileged access */
+#define DESCR_AP2     (1uL << 7)         /* Read only */
+#define DESCR_nSH     (0uL << 8)         /* Non-shareable */
```
I get the idea, but let's stick to the all-upper-case defines

agkaminski commented on hal/aarch64/zynqmp/interrupts.c:
```diff
@@ -0,0 +1,273 @@
+} intr_handler_t;
+
+
+struct {
```
```suggestion
static struct {
```

agkaminski commented on hal/aarch64/zynqmp/zynqmp.c:
```diff
@@ -0,0 +1,929 @@
+
+extern int _zynqmp_ddrInit(void);
+
+struct {
```
```suggestion
static struct {
```

agkaminski reviewed (commented):
General remarks:
- some file headers need changes (rarely the description, mostly missing author and year)
- it is nice that we can share plo device drivers between US+ and 7000, but these should be renamed to reflect that (perhaps [dev]-zynq?). Perhaps we could reduce number of ifdefs by extracting the platform dependent part to some other place

jmaksymowicz commented:
>     * it is nice that we can share plo device drivers between US+ and 7000, but these should be renamed to reflect that (perhaps [dev]-zynq?). Perhaps we could reduce number of ifdefs by extracting the platform dependent part to some other place

I renamed the shared device drivers to *-zynq and also added two words incorrectly picked up by spellchecker to `.codespell_ignore`
